File: collect_app/src/main/java/org/odk/collect/android/utilities/MultiClickGuard.java
package org.odk.collect.android.utilities;

import android.os.SystemClock;

public final class MultiClickGuard {
    private static final int CLICK_DEBOUNCE_MS = 1000;
    public static boolean test;

    private static long lastClickTime;
    private static String lastClickName;

    private MultiClickGuard() {
    }

    // Debounce multiple clicks within the same screen
    public static boolean allowClick(String className) {
        if (test) {
            return true;
        }

        long elapsedRealtime = SystemClock.elapsedRealtime();
        boolean isSameClass = className.equals(lastClickName);
        boolean isBeyondThreshold = elapsedRealtime - lastClickTime > CLICK_DEBOUNCE_MS;
        boolean isBeyondTestThreshold = lastClickTime == 0 || lastClickTime == elapsedRealtime; // just for tests
        boolean allowClick = !isSameClass || isBeyondThreshold || isBeyondTestThreshold;
        if (allowClick) {
            lastClickTime = elapsedRealtime;
            lastClickName = className;
        }
        return allowClick;
    }
}

File: collect_app/src/main/java/org/odk/collect/android/utilities/MultiClickGuard.kt
package org.odk.collect.android.utilities

import android.os.SystemClock

object MultiClickGuard {
    private const val CLICK_DEBOUNCE_MS = 1000

    @JvmField
    var test = false

    private var lastClickTime: Long = 0
    private var lastClickName: String = javaClass.name

    // Debounce multiple clicks within the same screen
    @JvmStatic
    fun allowClick(className: String = javaClass.name): Boolean {
        if (test) {
            return true
        }
        val elapsedRealtime = SystemClock.elapsedRealtime()
        val isSameClass = className == lastClickName
        val isBeyondThreshold = elapsedRealtime - lastClickTime > CLICK_DEBOUNCE_MS
        val isBeyondTestThreshold =
            lastClickTime == 0L || lastClickTime == elapsedRealtime // just for tests

        val allowClick = !isSameClass || isBeyondThreshold || isBeyondTestThreshold

        if (allowClick) {
            lastClickTime = elapsedRealtime
            lastClickName = className
        }
        return allowClick
    }
}
